### ISSUE TITLE
Require /login transactions to be executed in callback URL

### ISSUE THREAD

**Issue by kcole16** [near/near-wallet#726]
### Overview
The fix in #719 provides the users with more info to prevent the below vulnerability. To further reduce the likelihood of phishing, we can move the execution of the `addKey` TX on `/login` out of the wallet, and instead return the signed TX to the URL in the `callback_url`. 

Steps to expose vulnerability:
1. I wish to gain access to Kendall's Flux account.
2. I initiate a signin to Flux. At this point a new keypair K is generated in my session, and the private key is stored in my local storage. I am directed to a URL that will ask me for permission, and that URL contains the public key of that keypair.
3. I send that URL to Kendall. When he opens it, the screen legitimately asks for access to Flux, and is from near.org domain. Kendall has no reason not to trust it and not to Allow access.
4. Once Kendall allows access, the public key is added to his account, while the private key is in my local storage. I have full access to his Flux account.

### Acceptance Criteria
- [ ] If `callback_url` isn't valid, display error "This app cannot be authorized. This may be due to a problem with the app, or can be an attempt at phishing. We do not recommend using the app at this time."
- [ ] After user clicks "Allow", user is redirected to `callback_url` with `signed_tx` as an argument

**Comment by kcole16** [comment]
Note that this likely adds a bit of work for apps when it comes to handling and communicating TX errors to the user. They should be doing this already anyway, but can effectively offload some of this to the wallet in the current flow.

**Comment by vgrichina** [comment]
@kcole16 yeah, biggest problem I can see is that technically it's possible to use up same nonce already before app gets chance to do it. However thankfully it's pretty unlikely in normal usage. Like you'll have to login into other app (through same wallet) or sign transaction from app (through same wallet) while app is submitting transaction from existing login flow. 

So it's likely just ok to have generic "something went wrong, please login again" flow on the receiving side.

**Comment by kcole16** [comment]
From @vgrichina 
**Problem**

Referrer not always reliable as suggested by @abacabadabacaba:

Many websites allow users to post arbitrary link, and these links usually don't block the referrer. This is generally not considered a security issue, unless the page's URL is secret.
Likewise, many websites have redirectors that can redirect to an arbitrary URL. This is also not considered to be problematic by itself, even by Google.
Referrer may be blocked in some browsers.

**Proposed Solution**

Callback URL should be used as primary source of truth (especially after #726 lands). However if Referrer is set and isn't matching callback URL – we should report it as might be a sign of something sketchy going on.

**Comment by vgrichina** [comment]
Just realized that the proposed flow also fails with 2FA (and same for signing requests).

Wonder what is the best way to resolve this.

Maybe we should change the 2FA flow to be "contract helper creates multisig request, wallet confirms?". Wonder what your thoughts on that @mattlockyer @abacabadabacaba? Maybe you have some other ideas how to solve this issue however.

**Comment by abacabadabacaba** [comment]
One approach is to include a signed transaction from the contract helper as well. Or to implement multiparty signature generation so that the user and the contract helper jointly sign the same transaction.
